pkgs/unified_analytics/lib/src/event.dart

```diff
@@ -716,11 +719,89 @@ final class Event {
       other is Event &&
       other.runtimeType == runtimeType &&
       other.eventName == eventName &&
-      compareEventData(other.eventData, eventData);
-
-  @override
-  String toString() => jsonEncode({
+      _compareEventData(other.eventData, eventData);
+
+  /// Converts an instance of [Event] to JSON.
+  ///
+  /// Example for [Event.timing] converted to JSON below.
+  /// ```json
+  /// {
+  ///   "eventName": "timing",
+  ///   "eventData": {
+  ///     "workflow": "my-work-flow",
+  ///     "variableName": "my-variable",
+  ///     "elapsedMilliseconds": 123,
+  ///     "label": "my-label"
+  ///   }
+  /// }
+  /// ```
+  String toJson() => jsonEncode({
         'eventName': eventName.label,
         'eventData': eventData,
       });
+
+  @override
+  String toString() => toJson();
+
+  /// Utility function to take in two maps [a] and [b] and compares them
+  /// to ensure that they have the same keys and values.
+  ///
+  /// Used for the equality operator.
+  bool _compareEventData(Map<String, Object?> a, Map<String, Object?> b) {
+    final keySetA = a.keys.toSet();
+    final keySetB = b.keys.toSet();
+    final intersection = keySetA.intersection(keySetB);
+
+    // Ensure that the keys are the same for each object
+    if (intersection.length != keySetA.length ||
+        intersection.length != keySetB.length) {
+      return false;
+    }
+
+    // Ensure that each of the key's values are the same
+    for (final key in a.keys) {
+      if (a[key] != b[key]) return false;
+    }
+
+    return true;
+  }
+
+  /// Returns a valid instance of [Event] if [json] follows the correct schema.
+  ///
+  /// Schema is below with the following notes:
+  /// - The `eventName` key must have a string value that exists in [DashEvent].
+  /// - The `eventData` key must have a nested object as its value.
+  /// ```json
+  /// {
+  ///   "eventName": "string-label",
+  ///   "eventData": {
+  ///     "myVar1": "value1",
+  ///     "myVar2": 123
+  ///   }
+  /// }
+  /// ```
+  static Event? fromJson(String json) {
+    try {
+      final jsonMap = jsonDecode(json) as Map<String, Object?>;
+
+      if (!jsonMap.containsKey('eventName') ||
+          !jsonMap.containsKey('eventData')) {
+        return null;
+      }
+
+      final dashEvent =
+          DashEvent.getDashEventByLabel(jsonMap['eventName'] as String);
+      if (dashEvent == null) {
+        return null;
+      }
+      final eventData = jsonMap['eventData'] as Map<String, Object?>;
+
+      return Event._(eventName: dashEvent, eventData: eventData);
+    } on FormatException {
+      return null;
+      // ignore: avoid_catching_errors
+    } on TypeError {
+      return null;
```

Comment:
Is catching a `TypeError` here still okay? This would catch any cases `eventName` isn't a string or `eventData` is not a nested map object.

Example below would throw since `eventData` is a string type
```json
{
  "eventName": "string-label",
  "eventData": "this-should-be-a-map"
}
```

Reply:
Actually it may be better to just conditionally check for type errors instead of catching them

```dart
      // Ensure the values for each key is the correct type
      final eventName = jsonMap['eventName'];
      final eventData = jsonMap['eventData'];
      if (eventName is! String || eventData is! Map<String, Object?>) {
        return null;
      }
```

Reply:
SGTM